### PR TITLE
squid: a series of optimizations for kerneldevice discard

### PR DESCRIPTION
**Pull request by YiteGu**
1. blk: mul-thread discard support
2. NCB fix for leaked space when bdev_async_discard is enabled
3. React to bdev_enable_discard changes in handle_conf_change(); Fix several issues with stopping discard threads

backport tracker: https://tracker.ceph.com/issues/67139
backport tracker: https://tracker.ceph.com/issues/67404

---

backport of https://github.com/ceph/ceph/pull/55469, https://github.com/ceph/ceph/pull/56744, https://github.com/ceph/ceph/pull/58409
parent tracker: https://tracker.ceph.com/issues/65298, https://tracker.ceph.com/issues/66817

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by YiteGu** [comment]
jenkins test make check

**Review by ifed01 — APPROVED** [review]

**Comment by ljflores** [comment]
@YiteGu @ifed01 please see this error found in QA testing that looks related to this PR:
/a/skanta-2024-09-27_06:56:34-rados-wip-bharath14-testing-2024-09-26-2119-squid-distro-default-smithi/7921618
```
2024-09-27T09:22:38.858 INFO:tasks.rados.rados.0.smithi006.stdout:4227:  finishing copy_from to smithi00633089-523
2024-09-27T09:22:38.858 INFO:tasks.rados.rados.0.smithi006.stdout:update_object_version oid 523 v 1068 (ObjNum 1802 snap 167 seq_num 1802) dirty exists
2024-09-27T09:22:38.867 INFO:teuthology.orchestra.run.smithi179.stderr:error preparing db environment: (5) Input/output error
2024-09-27T09:22:38.870 DEBUG:teuthology.orchestra.run:got remote process result: 1
2024-09-27T09:22:38.872 INFO:tasks.thrashosds.thrasher:Traceback (most recent call last):
  File "/home/teuthworker/src/github.com_ceph_ceph-c_396230dce0bbb423ab67ae5d3625348d61b9460c/qa/tasks/ceph_manager.py", line 192, in wrapper
    return func(self)
  File "/home/teuthworker/src/github.com_ceph_ceph-c_396230dce0bbb423ab67ae5d3625348d61b9460c/qa/tasks/ceph_manager.py", line 1435, in _do_thrash
    self.choose_action()()
  File "/home/teuthworker/src/github.com_ceph_ceph-c_396230dce0bbb423ab67ae5d3625348d61b9460c/qa/tasks/ceph_manager.py", line 1171, in test_bluestore_reshard
    self.test_bluestore_reshard_action()
  File "/home/teuthworker/src/github.com_ceph_ceph-c_396230dce0bbb423ab67ae5d3625348d61b9460c/qa/tasks/ceph_manager.py", line 1151, in test_bluestore_reshard_action
    raise Exception("ceph-bluestore-tool resharding failed.")
Exception: ceph-bluestore-tool resharding failed.
```

Ref: https://tracker.ceph.com/issues/68294

**Comment by YiteGu** [comment]
> @YiteGu @ifed01 please see this error found in QA testing that looks related to this PR: /a/skanta-2024-09-27_06:56:34-rados-wip-bharath14-testing-2024-09-26-2119-squid-distro-default-smithi/7921618
> 
> ```
> 2024-09-27T09:22:38.858 INFO:tasks.rados.rados.0.smithi006.stdout:4227:  finishing copy_from to smithi00633089-523
> 2024-09-27T09:22:38.858 INFO:tasks.rados.rados.0.smithi006.stdout:update_object_version oid 523 v 1068 (ObjNum 1802 snap 167 seq_num 1802) dirty exists
> 2024-09-27T09:22:38.867 INFO:teuthology.orchestra.run.smithi179.stderr:error preparing db environment: (5) Input/output error
> 2024-09-27T09:22:38.870 DEBUG:teuthology.orchestra.run:got remote process result: 1
> 2024-09-27T09:22:38.872 INFO:tasks.thrashosds.thrasher:Traceback (most recent call last):
>   File "/home/teuthworker/src/github.com_ceph_ceph-c_396230dce0bbb423ab67ae5d3625348d61b9460c/qa/tasks/ceph_manager.py", line 192, in wrapper
>     return func(self)
>   File "/home/teuthworker/src/github.com_ceph_ceph-c_396230dce0bbb423ab67ae5d3625348d61b9460c/qa/tasks/ceph_manager.py", line 1435, in _do_thrash
>     self.choose_action()()
>   File "/home/teuthworker/src/github.com_ceph_ceph-c_396230dce0bbb423ab67ae5d3625348d61b9460c/qa/tasks/ceph_manager.py", line 1171, in test_bluestore_reshard
>     self.test_bluestore_reshard_action()
>   File "/home/teuthworker/src/github.com_ceph_ceph-c_396230dce0bbb423ab67ae5d3625348d61b9460c/qa/tasks/ceph_manager.py", line 1151, in test_bluestore_reshard_action
>     raise Exception("ceph-bluestore-tool resharding failed.")
> Exception: ceph-bluestore-tool resharding failed.
> ```
> 
> Ref: https://tracker.ceph.com/issues/68294

At first glance, this backport does not modify the code related to ceph-bluestore-tool rehard action. where I can see log file of ceph-bluestore-tool rehard?

File: src/blk/kernel/KernelDevice.cc
Comment (ifed01) [inline src/blk/kernel/KernelDevice.cc:69]: One should include https://github.com/ceph/ceph/pull/59026 or leave discard_stop initialization to avoid valgrid errors

**Comment by YiteGu** [comment]
> Please bring discard_stop initialization back.

done

**Comment by YiteGu** [comment]
jenkins test make check

**Review by ifed01 — APPROVED** [review]

**Comment by ifed01** [comment]
> The second one "ceph-bluestore-tool resharding failed" and it needs further investigation. Not sure if it's related to this PR.

So after checking the logs and talking to Adam I'm pretty sure that my hypothesis is valid - there were two independent issues with the last QA run. The second one is tracked by https://tracker.ceph.com/issues/67911

**Comment by aclamk** [comment]
@ljflores The issue in https://github.com/ceph/ceph/pull/59065#issuecomment-2384029349 is caused by:
https://tracker.ceph.com/issues/67911

**Comment by ifed01** [comment]
So @ljflores - please include https://github.com/ceph/ceph/pull/59969 along with this PR to get rid of the above issues.

**Comment by ljflores** [comment]
@YiteGu @ifed01 This PR was merged prematurely before it got tested. Can you please review the revert PR (#60641) and re-raise this PR?

**Comment by parth-gr** [comment]
@ljflores  @SrinivasaBharath This might be the reason for Bluestore osd-exapnsion failing https://github.com/rook/rook/pull/15251#issuecomment-2582895014|

cc @travisn

**Comment by travisn** [comment]
Looks like the revert PR #60641 has not been merged yet. So if this is related, we should be able to see the latest squid devel image failing in the resize testing: `quay.ceph.io/ceph-ci/ceph:squid`

**Comment by parth-gr** [comment]
Yes its failing with `quay.ceph.io/ceph-ci/ceph:squid`

**Comment by solidDoWant** [comment]
Hey Ceph maintainers, [this change removed `bdev_async_discard`](https://github.com/ceph/ceph/pull/59065/files#diff-24521e8b04f166f7cfcd9bccf49c69817e42a1681fa98f1a8af11893413adedcL4022), which is a breaking change.

While this is mentioned [in the changelog](https://docs.ceph.com/en/latest/releases/squid/#v19-2-1-squid), it only includes the PR title which isn't very descriptive. Can this change be rolled back until v20, or be made compatible with the current `bdev_async_discard` field, or can a large warning be added to the v19.2.1 changelog?

**Comment by solidDoWant** [comment]
This change causing a pretty major performance regression when `bdev_async_discard_threads` is greater than 1:
image

image

The charge from ~0 to ~12 is when I changed from 1 to 4. 

I'm not the only one seeing this. Here's a few screenshots from some other users:

image
Before setting it, after raising it to 4, and after setting it to 1

image
Before setting it, after raising it to 4, and after dropping to 2

I can't file an issue on the official bugtracker because account creation requires manual approval. @YiteGu @ifed01 Sorry for the ping but I don't know how else to get this in front of somebody.

**Comment by YiteGu** [comment]
> Hey Ceph maintainers, [this change removed `bdev_async_discard`](https://github.com/ceph/ceph/pull/59065/files#diff-24521e8b04f166f7cfcd9bccf49c69817e42a1681fa98f1a8af11893413adedcL4022), which is a breaking change.
> 
> While this is mentioned [in the changelog](https://docs.ceph.com/en/latest/releases/squid/#v19-2-1-squid), it only includes the PR title which isn't very descriptive. Can this change be rolled back until v20, or be made compatible with the current `bdev_async_discard` field, or can a large warning be added to the v19.2.1 changelog?

The purpose of `bdev_async_discard_threads` is to enable concurrent completion of discard op using more threads. If set to size of one is no change from the existing behaviour. At this point, the `bdev_async_discard` option becomes redundant.

**Comment by solidDoWant** [comment]
I completely get that the `bdev_async_discard` field is now redundant, and setting it to `1` preserves the current behavior. However, because this requires ceph admins to take action to preserve the current behavior, this is by definition a breaking change.

Unless I am missing something here, ceph admins who have previously set `ceph config set global bdev_async_discard 1`, or have a config file with
```ini
[global]
bdev_async_discard = 1
```

will suddenly, without any notice, have async discards disabled on their clusters.

Additionally, running `ceph config set global bdev_async_discard 1` now fails on v19.2.1 forward, which breaks automation and other tools that integrate with ceph, like Rook.

This change is breaking compatibility with v19 stable, and without backwards compatibility should have been delayed until v20.

**Comment by YiteGu** [comment]
> I completely get that the `bdev_async_discard` field is now redundant, and setting it to `1` preserves the current behavior. However, because this requires ceph admins to take action to preserve the current behavior, this is by definition a breaking change.
> 
> Unless I am missing something here, ceph admins who have previously set `ceph config set global bdev_async_discard 1`, or have a config file with
> 
> ```ini
> [global]
> bdev_async_discard = 1
> ```
> 
> will suddenly, without any notice, have async discards disabled on their clusters.
> 
> Additionally, running `ceph config set global bdev_async_discard 1` now fails on v19.2.1 forward, which breaks automation and other tools that integrate with ceph, like Rook.
> 
> This change is breaking compatibility with v19 stable, and without backwards compatibility should have been delayed until v20.

I completely understand how you feel. It is indeed a bit troublesome during the period of introducing new configurations and canceling old configurations.

**Comment by YiteGu** [comment]
> This change causing a pretty major performance regression when `bdev_async_discard_threads` is greater than 1: image
> 
> image
> 
> The charge from ~0 to ~12 is when I changed from 1 to 4.
> 
> I'm not the only one seeing this. Here's a few screenshots from some other users:
> 
> image Before setting it, after raising it to 4, and after setting it to 1
> 
> image Before setting it, after raising it to 4, and after dropping to 2
> 
> I can't file an issue on the official bugtracker because account creation requires manual approval. @YiteGu @ifed01 Sorry for the ping but I don't know how else to get this in front of somebody.

19.2.1 is missing https://github.com/ceph/ceph/pull/59529 , causing excessive CPU load during multi-threaded discard, please keep `bdev_async_discard_threads = 1` at 19.2.1

**Pull request merged**
